File: plugins/markdown/src/org/intellij/plugins/markdown/lang/psi/MarkdownPsiFactory.java
package org.intellij.plugins.markdown.lang.psi;

import com.intellij.extapi.psi.ASTWrapperPsiElement;
import com.intellij.lang.ASTNode;
import com.intellij.psi.PsiElement;
import com.intellij.psi.tree.IElementType;
import org.intellij.plugins.markdown.lang.MarkdownElementTypes;
import org.intellij.plugins.markdown.lang.MarkdownTokenTypeSets;
import org.intellij.plugins.markdown.lang.psi.impl.*;
import org.jetbrains.annotations.NotNull;

public enum MarkdownPsiFactory {
  INSTANCE;

  public PsiElement createElement(@NotNull ASTNode node) {
    final IElementType elementType = node.getElementType();

    if (elementType == MarkdownElementTypes.PARAGRAPH) {
      return new MarkdownParagraphImpl(node);
    }
    if (MarkdownTokenTypeSets.HEADERS.contains(elementType)) {
      return new MarkdownHeaderImpl(node);
    }
    if (elementType == MarkdownElementTypes.CODE_FENCE) {
      return ((MarkdownCodeFenceImpl)node);
    }
    if (elementType == MarkdownElementTypes.IMAGE) {
      return new MarkdownImage(node);
    }
    if (MarkdownTokenTypeSets.LISTS.contains(elementType)) {
      return new MarkdownList(node);
    }
    if (elementType == MarkdownElementTypes.LIST_ITEM) {
      return new MarkdownListItem(node);
    }
    if (elementType == MarkdownElementTypes.BLOCK_QUOTE) {
      return new MarkdownBlockQuote(node);
    }
    if (elementType == MarkdownElementTypes.SHORT_REFERENCE_LINK) {
      return new MarkdownShortReferenceLink(node);
    }
    if (elementType == MarkdownElementTypes.LINK_DEFINITION) {
      return new MarkdownLinkDefinition(node);
    }
    if (elementType == MarkdownElementTypes.LINK_DESTINATION) {
      return new MarkdownLinkDestinationImpl(node);
    }
    if (elementType == MarkdownElementTypes.LINK_LABEL) {
      return new MarkdownLinkLabel(node);
    }
    if (elementType == MarkdownElementTypes.CODE_BLOCK) {
      return new MarkdownCodeBlockImpl(node);
    }
    if (elementType == MarkdownElementTypes.TABLE) {
      return new MarkdownTable(node);
    }
    if (elementType == MarkdownElementTypes.TABLE_ROW || elementType == MarkdownElementTypes.TABLE_HEADER) {
      return new MarkdownTableRow(node);
    }
    if (elementType == MarkdownElementTypes.TABLE_CELL) {
      return new MarkdownTableCell(node);
    }

    return new ASTWrapperPsiElement(node);
  }
}

File: plugins/markdown/core/src/org/intellij/plugins/markdown/lang/psi/MarkdownPsiFactory.kt
package org.intellij.plugins.markdown.lang.psi

import com.intellij.extapi.psi.ASTWrapperPsiElement
import com.intellij.lang.ASTNode
import com.intellij.psi.PsiElement
import org.intellij.plugins.markdown.lang.MarkdownElementTypes
import org.intellij.plugins.markdown.lang.MarkdownTokenTypeSets
import org.intellij.plugins.markdown.lang.psi.impl.MarkdownBlockQuote
import org.intellij.plugins.markdown.lang.psi.impl.MarkdownCodeBlock
import org.intellij.plugins.markdown.lang.psi.impl.MarkdownCodeFence
import org.intellij.plugins.markdown.lang.psi.impl.MarkdownComment
import org.intellij.plugins.markdown.lang.psi.impl.MarkdownFrontMatterHeader
import org.intellij.plugins.markdown.lang.psi.impl.MarkdownHeader
import org.intellij.plugins.markdown.lang.psi.impl.MarkdownHeaderContent
import org.intellij.plugins.markdown.lang.psi.impl.MarkdownImage
import org.intellij.plugins.markdown.lang.psi.impl.MarkdownInlineLink
import org.intellij.plugins.markdown.lang.psi.impl.MarkdownLinkDefinition
import org.intellij.plugins.markdown.lang.psi.impl.MarkdownLinkDestination
import org.intellij.plugins.markdown.lang.psi.impl.MarkdownLinkLabel
import org.intellij.plugins.markdown.lang.psi.impl.MarkdownLinkText
import org.intellij.plugins.markdown.lang.psi.impl.MarkdownList
import org.intellij.plugins.markdown.lang.psi.impl.MarkdownListItem
import org.intellij.plugins.markdown.lang.psi.impl.MarkdownParagraph
import org.intellij.plugins.markdown.lang.psi.impl.MarkdownShortReferenceLink
import org.intellij.plugins.markdown.lang.psi.impl.MarkdownTable
import org.intellij.plugins.markdown.lang.psi.impl.MarkdownTableCell
import org.intellij.plugins.markdown.lang.psi.impl.MarkdownTableRow
import org.intellij.plugins.markdown.lang.psi.impl.MarkdownWrappedAutoLink

object MarkdownPsiFactory {
  @JvmStatic
  fun createElement(node: ASTNode): PsiElement {
    return when (val elementType = node.elementType) {
      MarkdownElementTypes.PARAGRAPH -> MarkdownParagraph(node)
      MarkdownElementTypes.CODE_FENCE -> node as MarkdownCodeFence
      MarkdownElementTypes.FRONT_MATTER_HEADER -> node as MarkdownFrontMatterHeader
      MarkdownElementTypes.INLINE_LINK -> MarkdownInlineLink(node)
      MarkdownElementTypes.LINK_TEXT -> MarkdownLinkText(node)
      MarkdownElementTypes.IMAGE -> MarkdownImage(node)
      MarkdownElementTypes.LIST_ITEM -> MarkdownListItem(node)
      MarkdownElementTypes.BLOCK_QUOTE -> MarkdownBlockQuote(node)
      MarkdownElementTypes.SHORT_REFERENCE_LINK -> MarkdownShortReferenceLink(node)
      MarkdownElementTypes.LINK_DEFINITION -> MarkdownLinkDefinition(node)
      MarkdownElementTypes.LINK_DESTINATION -> MarkdownLinkDestination(node)
      MarkdownElementTypes.LINK_LABEL -> MarkdownLinkLabel(node)
      MarkdownElementTypes.CODE_BLOCK -> MarkdownCodeBlock(node)
      MarkdownElementTypes.TABLE -> MarkdownTable(node)
      MarkdownElementTypes.TABLE_ROW, MarkdownElementTypes.TABLE_HEADER -> MarkdownTableRow(node)
      MarkdownElementTypes.TABLE_CELL -> MarkdownTableCell(node)
      MarkdownElementTypes.AUTOLINK -> MarkdownWrappedAutoLink(node)
      MarkdownElementTypes.LINK_COMMENT -> MarkdownComment(node)
      else -> when {
        elementType in MarkdownTokenTypeSets.HEADER_CONTENT -> MarkdownHeaderContent(node)
        MarkdownTokenTypeSets.HEADERS.contains(elementType) -> MarkdownHeader(node)
        MarkdownTokenTypeSets.LISTS.contains(elementType) -> MarkdownList(node)
        else -> ASTWrapperPsiElement(node)
      }
    }
  }
}
